---
title: "PAX9 Variants Random Forest Model"
author: "Cara Yijin Zou"
date: '2020-05-08'
tags: ["Model", "PAX9","Dental","Oligodontia","Random Forest", "Cross Validation","Bioinformatics", "Machine Learning","Supervised"]
categories: ["Dental","Bioinformatics"]
output: html_document
summary: "Pax9 is a gene belonging to the family of paired-box genes that encode transcription factors involved in organogenesis. Defects in Pax9 can lead to various types of cancer as well as a condition known as oligodontia which causes missing teeth in individuals. A random forest model was used to predict the phenotypic effect of variants of PAX9."

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning=FALSE, error=FALSE,message = FALSE, eval = TRUE)
```

## Predicting the Effect of PAX9 variants using a Random Forest Model 

### Abstract
Pax9 is a gene belonging to the family of paired-box genes that encode transcription factors involved in organogenesis. Defects in Pax9 can lead to various types of cancer as well as a condition known as oligodontia which causes missing teeth in individuals. A random forest model was used to predict the phenotypic effect of variants of PAX9. Features that were used include: location of mutation, type of mutation, CADD score, protein change (BLOSUM80 score), and in-frame/out of frame mutation. A 10-fold cross validation was conducted to ensure that the model was not overfitting. The resulting random forest model had an 84.0% prediction accuracy, in which the sensitivity was 89.5%, the specificity was 90.9%, and the precision was 89.5%. However, due to the small size of the available data there are many limitations to the model and involving other features such as other gene interactions would be a major direction of future research on the PAX9 gene. 

### Introduction/Background
The PAX9 gene belongs to the family of paired box genes (PAX), which are highly conserved across species. It is located on chromosome 14 and codes for a transcription factor which plays a key role in fetal development. Mutations in the PAX9 gene can lead to different types of cancers (especially lung cancer) as well as oligodontia, a congenital defect of more than 6 missing teeth, especially the molars.The members of the PAX gene family usually consist of a paired-type homeodomain, an octapeptide and a paired box domain. The paired box domain is responsible for binding to DNA to facilitate transcription and is one of our major interests to examine whether a mutation in the gene encoding this structure can be an indication of potential cancer or tooth agenesis phenotype. There are five exons that make up the gene transcript for PAX9. The two most important exons for our project include exon 2 (37130902-37131298), which contains four nucleotides including the start codon, as well as exon 3 (37132102-37132728) which contains the majority of the paired box domain.

The features used by the random forest model includes: location of mutation, type of mutation, CADD score, protein change (BLOSUM80 score), and in-frame/out of frame mutation. The location of the mutation was important because preliminary analysis using a chi-squared test determined that the location of the mutation (in the paired box region or not) is not independent to whether an individual exhibits oligodontia. Additionally, many of the features were chosen based on known biological information of genes/proteins. Type of mutation and in frame/out of frame mutation were chosen because synonymous mutations are likely to have little deleterious effect while missense and nonsense mutations are likely to cause a more significant effect on the gene product. Insertions and deletions are likely to cause the most effect on the gene since this can cause a frameshift and thus affect all subsequent amino acids. The effect of the protein change was quantified using a higher valued BLOSUM (BLOSUM80) which is used for closely related sequences as some amino acid changes are more detrimental than others. Finally, CADD score was used as it has a similar objective to our model (predicting the effect of variants) and including other similar models will hopefully improve the ability of our model. 

### Objective 
Utilizing a random forest to classify whether a PAX9 gene variant is pathogenic with the gene data from ClinVar, cBioPortal and gnomAD. 

### Data Sources
The pathogenic variants were found using cBioPortal (https://www.cbioportal.org/) which contained known variants that cause cancer. Pathogenic variants were also identified from the ClinVar database (https://www.ncbi.nlm.nih.gov/clinvar/) which consists of variants and their corresponding phenotype/condition. ClinVar largely contained variants that resulted in tooth agenesis. If the clinical significance was labelled as “benign” or “likely benign”, they were excluded from the positive variants dataset. It must be noted that there is a risk using ClinVar since there are more lenient standards for submitting data onto the database. 

The negative variants were extracted from gnomAD (https://gnomad.broadinstitute.org/) which is an aggregation of genome sequencing data. Although variants from individuals that are affected by severe pediatric disease as well as their first degree relatives were removed, there may still be some individuals affected by severe disease still included. Therefore, this is not a true reference set of alleles. To remove some known disease variants, any overlap between gnomAD and the pathogenic variants were removed from the gnomAD dataset provided that the allele frequency was less than 0.05.  

All variants were relative to the GRCh37 reference genome (NG_013357.1). 

### Computational Methods
Various R packages were used including ‘tidyverse’ and 'ggplot2' for manipulating the data as well as ‘peptider’ which contained the BLOSUM80 matrix. The ‘randomForest’ package provided the model for machine learning. 

The Ensembl Variant Effect Predictor (VEP) was used to obtain the protein consequence for the ClinVar data while the other two datasets already included protein consequence. A VCF file was produced using the results of VEP and was used to obtain the PHRED score from Combined Annotation Dependent Depletion (https://cadd.gs.washington.edu/) which predicts the deleteriousness of a single nucleotide change as well as insertions and deletions. 

### Results

##### Variant Selection 
Since there are a lot more variants from gnomAD that are neutral phenotype compared to the number of pathogenic variants, an unbalanced data would form if all of the datasets were included. To combat this, we sampled an equal number of variants that are diseased/not diseased each time we trained our model. Additionally, we used the ten-fold cross validation to make sure that the data is not overfit. 

##### Features
Allele frequency was originally a feature, however, ClinVar did not contain allele frequency and any attempts at obtaining allele frequency using the Ensembl allele frequency calculator were not successful because it only retrieves information for variants identified from the 1000 Genomes Project. While the cBioPortal dataset did have allele frequency, this was the  frequency of a cancerous cell with the PAX9 variant out of all the biopsy samples which was not directly comparable to the gnomAD allele frequency value which is in reference to the entire population. Therefore, this feature was removed.
 
###### Gene Location 
The feature ‘gene location’ was derived from all the variant positions (only using the starting position for convenience) on the chromosome in reference to the human genome GRCh37. The classification of position was built up according to the UCSC genome browser with reference sequence NM_006194. Additionally, we enriched the classification result via the usage of the NCBI analysis on NM_006194.3 since USCS genome browser only provided the exon position. As a result, categories in this feature include: 

* 5’ UTR (37126773-37,131,294) (covering all the exon 1 and majority of exon 2)

* Start codon region (37131295-37131298) (at the very end of exon 2) 

* Exon 3 (37132102-37132728) (contain PAI, RED and KDM5B)

* PAI  (37132116-37132286)

* RED (37132341-37132487)

* KDM5B (37132599-37132664)

* Exon 4 (37135667-37135806)

* Exon 5 (37145403-37145657)

* 3’ UTR (37,145,658-37147011) 

* Intron (parts outside of the above regions) 

###### Paired box region 
The PAI stands for paired-box region, which encodes a structure for the PAX9 transcription factor to bind to DNA molecules. The RED subdomain aids and corrects PAI’s recognition of DNA sites, and its variant number is larger than PAI, which might account for why PAX proteins have different target genes. The KDM5B protein helps transcriptional repression of some tumor suppressor gene.   

###### BLOSUM80 (amino acid conservation)
The BLOSUM80 score displayed the extent of effect of amino acid change on the stability of the PAX9 protein. We expected that this feature would greatly improve the prediction of our model. R package ‘peptider’ was used to obtain the BLOSUM80 matrix. However, the substitution matrix only considered single amino acid change, so we assumed that variants with a replacement length over one (such as a deletion, frameshift, etc) would get the lowest score since they are more likely to have deleterious effects on the protein structure. 

###### Mutation Type
For the feature of mutation type, it was determined from the original data for gnomAD and cBioPortal. The ClinVar data was processed via VEP to acquire the result. Using the mutation feature and BLOSUM80 matrix, we were able to further classify the data into whether it was in-frameshift and out-of-frame shift since it was expected that frameshift mutation would likely to generate abnormal protein. 

###### CADD PHRED score 
CADD PHRED score was obtained by using the coordinates of variants and submitting to VEP. The result produced a VCF file that was submitted to CADD. Some variants such as fusion proteins did not have a CADD score and were given an estimated CADD score ( in this particular case, maximum CADD score was used in replacement) which may not have reflected its true deleteriousness effect.

###### Pathogenicity 
Finally, for the pathogenicity feature--which was our goal of RF prediction--due to the missing allele frequency in most of the cBioPortal and ClinVar it would be nearly impossible to precisely know which variant was neutral or not. Thus, we had to assume that all ClinVar and cBioPortal variants were pathogenic while all gnomAD variants were neutral (after filtering for duplicates) for convenience. 

```{r}
library(tidyverse)

clinvar<-read.csv("clinvar_result.csv", header=T,na.strings="")%>%mutate_each(as.character())
cancer<-read.csv("pax9_mutation_in_cancer.csv", header=T,na.strings="")
gnomad<-read.csv("gnomAD_Grch37.csv", header=T,na.strings="")

```

```{R}
#final few rows of clinvar do not have location of mutation, exclude those that are benign/likely benign
clinvar1<-clinvar%>%slice(1:57)%>%filter(grepl("benign", Clinical.significance..Last.reviewed.,fixed=T)!=T)%>%select(dbSNP.ID, GRCh37Location, Name, Protein.change, Clinical.significance..Last.reviewed.)


#turn into character instead of factor 
gnomAD1 = gnomad %>% select(rsID, Position, Consequence, Protein.Consequence, Transcript.Consequence, Annotation,Allele.Frequency, Reference, Alternate) %>%
  mutate(rsID = as.character(rsID)) %>% mutate(Consequence = as.character(Consequence)) %>% mutate(Protein.Consequence = as.character(Protein.Consequence)) %>%
  mutate(Transcript.Consequence = as.character(Transcript.Consequence)) %>% mutate(Annotation = as.character(Annotation))

#turn into character instead of factor 
cancer1 = cancer %>% select(dbSNP, Protein.Change, HGVSg, Start.Pos, Mutation.Type, Variant.Type,Ref,Var) %>% mutate(Protein.Change = as.character(Protein.Change)) %>%
  mutate(HGVSg = as.character(HGVSg)) %>% mutate(Mutation.Type = as.character(Mutation.Type)) %>% mutate(Variant.Type = as.character(Variant.Type)) %>%
  mutate(dbSNP = as.character(dbSNP))%>%mutate(Ref = as.character(Ref))%>%mutate(Var = as.character(Var))

```

```{R}
#no duplicates in positives
#inner_join(clinvar1,cancer1,by=c(dbSNP.ID="dbSNP"),na_matches="never")

#None of allele frequency over 0.05, therefore discard from gnomAD and keep in positive controls
#inner_join(gnomAD1,cancer1, by=c(rsID="dbSNP"),na_matches="never")%>%filter(Allele.Frequency>0.05)
#inner_join(gnomAD1,clinvar1, by=c(rsID="dbSNP.ID"),na_matches="never")%>%filter(Allele.Frequency>0.05)

#Before Duplicates
#gnomAD1%>%nrow()

gnomAD2=anti_join(gnomAD1,cancer1, by=c(rsID="dbSNP"),na_matches="never")
gnomAD2=anti_join(gnomAD2,clinvar1, by=c(rsID="dbSNP.ID"),na_matches="never")

#gnomAD2%>%nrow()
gnomAD2_copy=gnomAD2
#get rid of rsID in all datasets
clinvar1=clinvar1%>%select(-dbSNP.ID)
cancer1=cancer1%>%select(-dbSNP)
gnomAD2=gnomAD2%>%select(-rsID)

#Resulting datasets: clinvar1, gnomAD2, cancer1
```


```{R}
#Defining the Features

#Feature 1: Mutation Location
mutation_region = function(data){
  result = c()
  for(i in 1:length(data)){
	if(between(data[i], 37126773, 37131294) == TRUE){
  	result[i] = '5’ UTR'
  	}else if(between(data[i], 37131295, 37131298) == TRUE){
    	result[i] = 'Start codon region'
    	}else if(between(data[i], 37135667, 37135806) == TRUE){
      	result[i] = 'Exon 4'
      	}else if(between(data[i], 37145403, 37145657) == TRUE){
        	result[i] = 'Exon 5'
        	}else if(between(data[i], 37132102, 37132728) == TRUE){
          	if(between(data[i], 37132116, 37132286) == TRUE){
            	result[i] = 'PAI subdomain'
            	}else if(between(data[i], 37132341, 37132487) == TRUE){
              	result[i] = 'RED subdomain'
              	}else if(between(data[i], 37132599, 37132664) == TRUE){
                	result[i] = 'KDM5B'
                	}else(result[i] = 'Exon 3')
          	}else(result[i] ='Intron')
  }
  return(result)
}

```


```{R}
# Cancer Variant
cancer2 = cancer1 %>% mutate(gene_location = mutation_region(cancer1$Start.Pos))
# Clinvar Variant
clinvar2 = clinvar1 %>% mutate(gene_location = mutation_region(clinvar1$GRCh37Location))
# GnomAD Variant
gnomAD2 = gnomAD2 %>% mutate(gene_location = mutation_region(gnomAD2$Position))


```


```{R}
#Feature 2: BLOSUM score for amino acid conservativeness
library(peptider)
one_to_three=function(x){
  one<- c("A", "C", "D", "E", "F" ,"G", "H", "I", "K", "L", "M" ,"N", "P" ,"Q" ,"R", "S" ,"T", "V", "W", "Y")
  three<-c("Ala", "Cys" ,"Asp", "Glu" ,"Phe" ,"Gly" ,"His" ,"Ile", "Lys" ,"Leu" ,"Met","Asn", "Pro" ,"Gln" ,"Arg" ,"Ser" ,"Thr", "Val" ,"Trp", "Tyr")
  index=match(x,one)
  return(three[index])
}

#create BLOSUM with three letter amino acid code
BLOSUM80_2<-BLOSUM80%>%mutate(AA1=one_to_three(as.character(AA1)))%>%mutate(AA2=one_to_three(as.character(AA2)))

```

```{R}
##CANCER
select_cancer<-cancer%>%select(Protein.Change, HGVSg, Start.Pos, End.Pos, Ref, Var)
clean_cancer<-select_cancer%>%mutate(copyHGVSg = as.character(HGVSg))%>%separate(copyHGVSg, into=c(NA, NA, "temporary", "AA2"))%>%separate(temporary, into=c(NA, "AA1"), sep=8)
BLOSUM_score_cancer = left_join(clean_cancer, BLOSUM80,  by = c("AA1", "AA2"))%>%mutate(Blosum2=ifelse(is.na(Blosum)==T,-6, Blosum))
cancer2$BLOSUM = BLOSUM_score_cancer$Blosum2

##CLINVAR
select_clinvar<-clinvar2%>%select(Name)
clin_var_aa<-select_clinvar%>%mutate(copyName = as.character(Name))%>%separate(copyName, into=c(NA, "temp1"), sep=":")%>%mutate(five.UTR=ifelse(substr(temp1, 1, 3)=="c.-", "UTR", NA))%>%mutate(premature_stop=ifelse(substr(temp1, 1,3)=="c.*", "PREMATURE_STOP", NA))%>%separate(temp1, into=c(NA, NA, "temp3",NA,"Protein.Change"))%>%mutate(indel=ifelse(nchar(temp3)>1,"indel", NA))%>%mutate(AA1=substr(Protein.Change,1,3))%>%mutate(AA2=ifelse(nchar(Protein.Change)>6, substr(Protein.Change, nchar(Protein.Change)-2, nchar(Protein.Change)),AA1))%>%unite(Protein.Effect,c(Protein.Change, five.UTR, indel, premature_stop),na.rm=TRUE, remove=TRUE)
BLOSUM_score_clinvar = left_join(clin_var_aa, BLOSUM80_2,  by = c("AA1","AA2"))%>%mutate(Blosum=ifelse(is.na(Blosum),ifelse(startsWith(Protein.Effect,"indel")|startsWith(Protein.Effect,"PREMATURE_STOP"),-6,0),Blosum))
clinvar2$BLOSUM = BLOSUM_score_clinvar$Blosum

##GNOMAD
protein_effect=function(x){
  cleanx=substr(x,3,nchar(x))
  start=substr(cleanx,1,3)
  end=substr(gsub('[0-9_]+', '', substr(cleanx,4,nchar(cleanx))),1,3)
  return (paste(start,end, sep="_"))
}

gnomad_aa<-gnomAD2%>%select(Annotation,Consequence)%>%mutate(Consequence=as.character(Consequence))%>%mutate(result=ifelse(Annotation=="5_prime_UTR_variant"|Annotation=="intron_variant"|Annotation=="splice_region_variant"|Annotation=="3_prime_UTR_variant" ,NA,protein_effect(Consequence)))%>%separate(result, into=c("AA1","AA2"))
BLOSUM_score_gnomAD = left_join(gnomad_aa, BLOSUM80_2, by=c("AA1","AA2"))%>%mutate(Blosum=ifelse(is.na(Blosum),ifelse(AA2=="Ter"|AA2=="",-6,0),Blosum))%>%mutate(Blosum=ifelse(is.na(Blosum),0,Blosum))
gnomAD2$BLOSUM = BLOSUM_score_gnomAD$Blosum

```

```{R}
#Feature 3: Mutation Type 
#The mutation type was called different names in the dataset. This step was used to make this feature uniform and extract the mutation type from each of the datasets. 
## cancer variant
Mutation.Type <- c("Splice_Region",'Missense_Mutation','Nonsense_Mutation','Frame_Shift_Del',"Fusion","Splice_Site")
cancer_translation<-c("Splice", "Missense", "Nonsense", "Out_Frame_Del","Fusion", "Splice")
cancer_trans<- data.frame(Mutation.Type, cancer_translation)
cancer_type<-cancer%>%select(Mutation.Type)
Mutation_type_cancer = left_join(cancer_type, cancer_trans,by="Mutation.Type")
cancer2$Mutation_Type = as.character(Mutation_type_cancer$cancer_translation)



# gnomAD variant
Annotation<-c("5_prime_UTR_variant","intron_variant","splice_region_variant","missense_variant","synonymous_variant","stop_gained","inframe_deletion","inframe_insertion","start_lost","stop_lost","3_prime_UTR_variant")
gnomad_mutation_type<-c("5_prime_UTR_variant","Intron","Splice","Missense","Synonymous","Nonsense","inframe_deletion","inframe_insertion","start_lost","stop_lost","3_prime_UTR_variant")
gnomad_trans<-data.frame(Annotation, gnomad_mutation_type)
gnomad_type<-gnomAD2%>%select(Annotation)
Mutation_type_gnomAD = left_join(gnomad_type,gnomad_trans,by="Annotation")
gnomAD2$Mutation_Type = as.character(Mutation_type_gnomAD$gnomad_mutation_type)




#The original dataset for clinvar did not have this feature, therefore the coordinates of the mutation were used and inputted in VEP from ensemble. 

#clinvar variant
clinvar_VEP<-read.csv("Final_ClinVar_VEP.csv", header=T)
clinvar_VEP_select<-clinvar_VEP%>%select(X.Uploaded_variation,Consequence,Feature)%>%filter(Feature=="ENST00000361487.6")
Consequence<-c("upstream_gene_variant","5_prime_UTR_variant","non_coding_transcript_exon_variant","intron_variant,non_coding_transcript_variant","regulatory_region_variant","intron_variant","TF_binding_site_variant","downstream_gene_variant","start_lost,splice_region_variant","splice_region_variant,non_coding_transcript_exon_variant","splice_region_variant,synonymous_variant", "missense_variant", "synonymous_variant","stop_gained","frameshift_variant", "3_prime_UTR_variant" )
clin_var_translation<-c("5_prime_UTR_variant","5_prime_UTR_variant","5_prime_UTR_variant","Intron","regulatory_region_variant","Intron","TF_binding_site_variant", "3_prime_UTR_variant","start_lost","Splice","Splice","Missense","Synonymous","Nonsense","Out_Frame_Del","3_prime_UTR_variant")
clin_var_translation<-data.frame(Consequence, clin_var_translation)
Mutation_type_clinvar = left_join(clinvar_VEP_select,clin_var_translation)
clinvar2$Mutation_Type = as.character(Mutation_type_clinvar$clin_var_translation)

```

```{R}
#Feature 4: In frame or Out of frame Mutation 
#Clinvar
inframe_clinvar = left_join(clin_var_aa, BLOSUM80_2,  by = c("AA1","AA2"))%>%mutate(Blosum=ifelse(is.na(Blosum),ifelse(startsWith(Protein.Effect,"indel")|startsWith(Protein.Effect,"PREMATURE_STOP"),-6,0),Blosum))%>%mutate(frame=ifelse(nchar(temp3)>1, "out_frame","in_frame"))
clinvar2$inframe = inframe_clinvar$frame

#Cancer
inframe_cancer = left_join(clean_cancer, BLOSUM80,  by = c("AA1", "AA2"))%>%mutate(Blosum2=ifelse(is.na(Blosum)==T,-6, Blosum))%>%mutate(Ref=as.character(Ref))%>%mutate(frame=ifelse(nchar(Ref)>1|AA1=="del","out_frame", "in_frame"))
cancer2$inframe = inframe_cancer$frame

#GnomAD
inframe_gnomAD = left_join(gnomad_aa, BLOSUM80_2, by=c("AA1","AA2"))%>%mutate(Blosum=ifelse(is.na(Blosum),ifelse(AA2=="Ter"|AA2=="",-6,0),Blosum))%>%mutate(Blosum=ifelse(is.na(Blosum),0,Blosum))%>%mutate(Temp=Consequence)%>%separate(Temp, into=c("test1","test2"),sep="d")%>%mutate(frame=ifelse(is.na(test2),"in_frame", "out_frame"))
gnomAD2$inframe = inframe_gnomAD$frame

```


```{R}
#Feature 5: CADD score
#The CADD score obtained by first obtaining a VCF file from VEP by inputting variant coordinates. This VCF file was then submitted to CADD. 
#Clinvar
CADD_clinvar = read.csv('Final_ClinVar_CADD.csv', header=T)
clinvar2$PHRED = CADD_clinvar$PHRED

#Cancer variant
CADD_cancer = read.csv('Final_Cancer_CADD.csv', header = T)%>%mutate(Ref=as.character(REF))%>%mutate(Var=as.character(ALT))%>%rename(Start.Pos=POS)%>%select(-c(REF,ALT))
CADD_cancer<-CADD_cancer%>%mutate(Start.Pos=ifelse(nchar(Ref)>1, Start.Pos+1,Start.Pos))%>%mutate(Var=ifelse(nchar(Ref)>1,ifelse(substr(Var,2,nchar(Var))=="","-",substr(Var,2,nchar(Var))),Var))%>%mutate(Ref=ifelse(nchar(Ref)>1,substr(Ref,2,nchar(Ref)),Ref))
cancer3<-left_join(cancer, CADD_cancer, by=c("Start.Pos","Ref","Var"))%>%mutate(PHRED=ifelse(is.na(PHRED),max(PHRED,na.rm=T),PHRED))
cancer2$PHRED <- cancer3$PHRED

# gnomAD variant
CADD_gnomAD = read.csv('GnomAD_3_CADD.csv', header = T)
CADD_gnomAD<-CADD_gnomAD%>%rename(Reference=REF)%>%rename(Alternate=ALT)%>%rename(Position=POS)


gnomad_condensed<-gnomAD2_copy %>% select(Position,Consequence,Transcript.Consequence,Annotation,rsID,Reference,Alternate)



gnomAD_CADD<-left_join(gnomad_condensed,CADD_gnomAD, by=c("Position","Reference","Alternate"))



#Get the average score based on Annotation
Avg_CADD_Gnomad<-gnomAD_CADD%>%group_by(Annotation)%>%summarize(avg=mean(PHRED, na.rm=T))
result = left_join(gnomAD_CADD,Avg_CADD_Gnomad)%>% mutate(PHRED=ifelse(is.na(PHRED),avg,PHRED))
index = is.nan(result$PHRED)
index = result[as.logical(-1*index+1),]
average = mean(index$PHRED)
result = result %>% mutate(PHRED = ifelse(is.nan(PHRED), average, PHRED ))

gnomAD2$PHRED = result$PHRED


```

```{R}
#Final Processing
gnomAD2$phenotype = 'neutral'
clinvar2$phenotype = 'pathogenic'
cancer2$phenotype = 'pathogenic'
name = c('GRCh37Location', 'gene_location', 'BLOSUM', 'Mutation_Type', 'inframe', 'PHRED', 'phenotype')

gnomAD2 = gnomAD2 %>% rename(GRCh37Location=Position)%>% select(GRCh37Location, gene_location, BLOSUM, Mutation_Type, inframe, PHRED, phenotype)
clinvar2 = clinvar2 %>% select(GRCh37Location, gene_location, BLOSUM, Mutation_Type, inframe, PHRED, phenotype)
cancer2 = cancer2 %>% rename(GRCh37Location=Start.Pos)%>% select(GRCh37Location, gene_location, BLOSUM, Mutation_Type, inframe, PHRED, phenotype)

colnames(gnomAD2) = name
colnames(cancer2) = name
gnomAD2 = gnomAD2 %>%mutate(GRCh37Location = as.numeric(GRCh37Location)) %>% mutate(gene_location = as.factor(gene_location)) %>%
  mutate(Mutation_Type = as.factor(Mutation_Type)) %>% mutate(inframe = as.factor(inframe)) %>% mutate(phenotype = as.factor(phenotype))
clinvar2 = clinvar2 %>%mutate(GRCh37Location = as.numeric(GRCh37Location)) %>% mutate(gene_location = as.factor(gene_location)) %>%
  mutate(Mutation_Type = as.factor(Mutation_Type)) %>% mutate(inframe = as.factor(inframe)) %>% mutate(phenotype = as.factor(phenotype))
cancer2 = cancer2 %>%mutate(GRCh37Location = as.numeric(GRCh37Location)) %>% mutate(gene_location = as.factor(gene_location)) %>%
  mutate(Mutation_Type = as.factor(Mutation_Type)) %>% mutate(inframe = as.factor(inframe)) %>% mutate(phenotype = as.factor(phenotype))

```

##### Random Forest

```{R}
library(randomForest)
set.seed(50)
options(expressions = 5e5)
#pos/neg control
neg_control = rbind(clinvar2, cancer2) #112 pathogenic variants
pos_control_index = sample(nrow(gnomAD2), size = nrow(neg_control), replace = FALSE)
pos_control = gnomAD2[pos_control_index,]
# take a fifth of each data set for testing data
neg_control_index = sample(nrow(neg_control), size = round(nrow(neg_control)*0.2), replace = FALSE)
pos_control_index1 = sample(nrow(pos_control), size = round(nrow(pos_control)*0.2), replace = FALSE)
neg_testing_data = neg_control[neg_control_index,]
neg_training_data = neg_control[-neg_control_index,]
 
pos_testing_data = pos_control[pos_control_index1,]
pos_training_data = pos_control[-pos_control_index1,]
# 10-fold cross validation
 
training_data = rbind(pos_training_data, neg_training_data)
training_data = training_data[sample(nrow(training_data), size = nrow(training_data), replace = FALSE),]

ten_fold_accuracy = c()
for(i in 1:10){
  Index = (1:18)+(i-1)*18
  train_data = training_data[-Index, ]
  testin_data = training_data[Index, ]
  cross_result = randomForest(phenotype ~., data = train_data, importance = T, ntree = 50,
                               	na.action = na.exclude)
  result = predict(cross_result, testin_data)
 
  ten_fold_accuracy[i] = sum(result == testin_data$phenotype)/nrow(testin_data)
}

testin_data = training_data[145:162,]
train_data = training_data[-(145:162),]
cross_result = randomForest(phenotype ~., data = train_data, importance = T, ntree = 50,
                        	na.action = na.exclude)
result = predict(cross_result, testin_data)

tenfoldcv_result=matrix(ten_fold_accuracy, nrow = 1, dimnames = list('ten_fold_accuracy'))
tenfold_mean= mean(ten_fold_accuracy)
```

```{R}
# random forest
testin_data = rbind(neg_testing_data, pos_testing_data)
RF1 = randomForest(phenotype ~., data = training_data, importance = T, ntree = 20,
                             	na.action = na.exclude)
prediction_result = predict(RF1, testin_data )
RF1 
```
The RF model showed that it had an overall accuracy of 85.56%, with 20 decision trees for doing classification tasks. The low value of tree number was due to the limited training data size (80).

```{R}
importance (RF1)
varImpPlot(RF1)

```
Based on the variable importance plot above, the most important variable would be the location relative to the reference GRCh37. On the other hand, the variable with the smallest effect on the predicting power of our model is inframe. This is expected since there were relatively few variants that caused a frame shift in the population. 

##### Ten-fold Cross Validation 

```{R}
tenfold_df<-tenfoldcv_result%>%as.data.frame()
colnames(tenfold_df)<-c(1:10)
tenfold_df
```
Cross Validation mean accuracy:
```{R}
tenfold_mean
```
By comparing the accuracy between RF (85.56%) and ten-fold (86.67%), we found that despite the low sample size, the model exhibited limited overfitting.

##### Prediction Result
```{R}
# table of RF prediction 
cm = table(truth=testin_data$phenotype, prediction_result)
cm%>%addmargins()
accuracy=sum(cm[1,2],cm[2,1])/sum(cm)
precision=cm[1,2]/colSums(cm)[2]
sensitivity=cm[1,2]/colSums(cm)[2]
specificity=cm[2,1]/rowSums(cm)[2]
data.frame(accuracy,sensitivity,specificity,precision)

```
From the confusion matrix, the model has relatively good accuracy, sensitivity, specificity, and precision. 

### Conclusion
The objective of the project was to build a random forest model to predict the phenotype of a PAX9 gene variant. The ten-fold cross validation indicated that the model didn’t overfit. While the prediction result was satisfactory, it is important to note that it is only as good as the available training data. It is again acknowledged that there was a risk associated with using the ClinVar resource due to its relatively loose quality check. Additionally, since a large proportion of the allele frequency in cBioPortal and ClinVar was missing, it was difficult to identify all the potential neutral variants and all the variants were assumed to be pathogenic in order to run the RF model. Hence, if possible, allele frequency could potentially be an important feature of any future research on PAX9 gene. Another improvement would be to increase the very limited online human PAX9 gene data to generate more powerful machine learning models. To further refine the prediction of the RF model, a suggestion may be to use the continuous conservation score from other available tools as highly conserved sequences are likely less tolerant to changes. Although BLOSUM80 was insightful on the effect of amino acid substitution on protein structure, adding conservation score could provide a more comprehensive information. Lastly, future research may have a stronger focus on the PAX9’s metabolic pathway. This gene plays a principal role in organogenesis and interacts with many other genes so it is unfortunate that this gene hasn’t received the same attention as other popular genes like p53. 


### References
1. Bonczek et al, ‘PAX9 gene mutations and tooth agenesis: A review’, Clinical Genetics, 2017/02/01, doi: 10.1111/cge.12986.
https://www.researchgate.net/publication/313264549_PAX9_gene_mutations_and_tooth_agenesis_A_review
2. UCSC genome browser on PAX9. http://www.genome.ucsc.edu/cgi-bin/hgc?hgsid=828585015_qsdofxT4g0z4a48R33zLnyqdtxND&g=htcCdnaAliInWindow&i=NM_006194&c=chr14&l=37126772&r=37147011&o=37126772&aliTable=refSeqAli&table=refGene
3. NCBI data on PAX9: https://www.ncbi.nlm.nih.gov/nuccore/NM_006194#feature_NM_006194.3
4.  Pellizzari et al. Co-operation between the PAI and RED subdomains of Pax-8 in the interaction with the thyroglobulin promoter. Biochem J. 1999;337 ( Pt 2)(Pt 2):253‐262. 
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1219959/
5. KDB5B gene: https://www.genecards.org/cgi-bin/carddisp.pl?gene=KDM5B
6. Ensembl Variant Effect Predictor (VEP): https://useast.ensembl.org/info/docs/tools/vep/index.html
7. Combined Annotation Dependent Depletion (CADD): https://cadd.gs.washington.edu/
8. cBioPortal: https://www.cbioportal.org/
9. ClinVar:https://www.ncbi.nlm.nih.gov/clinvar/
10. gnomAD: https://gnomad.broadinstitute.org/
11. Random Forest: A. Liaw and M. Wiener (2002). Classification and Regression by randomForest. R News 2(3), 18--22. 
12. Tidyr package: Hadley Wickham and Lionel Henry (2020). tidyr: Tidy Messy Data. R package;  version 1.0.2. https://CRAN.R-project.org/package=tidyr
13. Peptider package: Heike Hofmann, Eric Hare and GGobi Foundation (2015). peptider: Evaluation of Diversity in Nucleotide Libraries. R package version 0.2.2.  https://CRAN.R-project.org/package=peptider
14. Dplyr package: Hadley Wickham, Romain François, Lionel Henry and Kirill Müller (2020). Dplyr:  A Grammar of Data Manipulation. R package version 0.8.5. https://CRAN.R-project.org/package=dplyr

