---
title: "Multiple Sequence Alignment"
author: "Cara Yijin Zou"
date: '2021-07-05'
draft: false
tags: ["Bioinformatics", "Multiple Sequence Alignment", "Dental"]
categories: ["Bioinformatics"]
output: html_document
summary: "Finding missense variants that help explain the differences in dentition in Neanderthals and modern humans using a multiple sequence alignment (R seqinr and msa package)." 
image:
  placement: 1
  caption: "DMP1 Multiple Sequence Alignment"
  focal_point: "Center"
  preview_only: true
---  

<script src="{{< blogdown/postref >}}index_files/header-attrs/header-attrs.js"></script>


<p>The method and database information was from following journal article: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5597096/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5597096/</a>. The objective was to discover missense variants that could help explain the differences in dentition between Neanderthals and modern humans. This was an attempt to recreate/speculate what was done for the research paper to obtain their results.</p>
<div id="read-in-the-fastas" class="section level2">
<h2>Read in the Fastas</h2>
<p>The function read.fasta() from the R package ‘seqinr’ was used to read in protein sequence FASTA files.</p>
<pre class="r"><code>#read fasta
library(seqinr)

#multiple sequence alignment
library(msa)

#read in fasta using seqinr
altai &lt;- read.fasta(&quot;Altai.txt&quot;, seqtype=&quot;AA&quot;,as.string=TRUE)

vindija &lt;- read.fasta(&quot;Vindija.txt&quot;, seqtype=&quot;AA&quot;,as.string=TRUE)

sidron &lt;- read.fasta(&quot;Sidron.txt&quot;, seqtype=&quot;AA&quot;,as.string=TRUE)</code></pre>
</div>
<div id="create-a-function-that-takes-the-gene-name-and-outputs-a-msa" class="section level2">
<h2>Create a function that takes the gene name and outputs a msa</h2>
<div id="part-1-create-a-combined-fasta-using-write.fasta" class="section level3">
<h3>Part 1: Create a combined fasta using write.fasta</h3>
<ul>
<li>The reference FASTA files for human, chimpanzee (<em>Pan troglodytes</em>), and gorilla (<em>Gorilla gorilla</em>) protein sequences were obtained from UniProt</li>
<li>Neanderthal data including Altai pedal phalanx (Altai Mountain Cave, Siberia) and bone fragments of SD1253 (El Sidron cave, Spain) and Vi33.15 (Vindija cave, Croatia) were obtained from the Max Planck Institute for Evolutionary Anthropology: <a href="http://cdna.eva.mpg.de/neandertal/exomes/" class="uri">http://cdna.eva.mpg.de/neandertal/exomes/</a></li>
</ul>
</div>
<div id="part-2-align-the-sequences-from-the-combined-fasta-using-the-msa-package" class="section level3">
<h3>Part 2: Align the sequences from the combined fasta using the msa package</h3>
<p>The ClustalW algorithm was used for the multiple sequence alignment with a BLOSUM matrix. The msaPrettyPrint() function was used to print the sequence in a more readable manner</p>
<pre class="r"><code>msa_maker&lt;- function(gene_name, ensg_num, hcg_fasta_name, combined_fasta_name){
  #Part 1: Create a combined fasta
  #Read in fasta with human, chimp, and gorilla 
  hcg_fasta &lt;- read.fasta(hcg_fasta_name, seqtype=&quot;AA&quot;,as.string=TRUE,set.attributes = FALSE)
  
  #changed the original sequence names because they were too long and made the final msa look messy 
  hcg_names &lt;- sapply(strsplit(names(hcg_fasta), &quot;|&quot;, fixed=TRUE),&quot;[[&quot;, 3)
  #write into new combined fasta 
  write.fasta(hcg_fasta, names=hcg_names,file.out = combined_fasta_name, open=&quot;w&quot;, as.string=TRUE)

  #find the protein in the Neanderthal Data and write it into the combined fasta file 
  #Altai
  #sequence name in file is in the form of ENSGXXXX_ENSTXXXX, grep used to just search with the ENSG info
  altai_index&lt;-grep(ensg_num,attributes(altai)$names)[[1]] 
  altai_filtered&lt;- altai[altai_index]
  #write into new combined fasta
  write.fasta(paste(altai_filtered)[1], paste0(&quot;altai_&quot;,gene_name),file.out = combined_fasta_name, open=&quot;a&quot;, as.string=TRUE)
  
  #Vindija
  vindija_index&lt;-grep(ensg_num,attributes(vindija)$names)[[1]]
  vindija_filtered&lt;-vindija[vindija_index]
  write.fasta(paste(vindija_filtered)[1], paste0(&quot;vindija_&quot;, gene_name),file.out = combined_fasta_name, open=&quot;a&quot;, as.string=TRUE)
  
  #Sidron
  sidron_index&lt;-grep(ensg_num,attributes(sidron)$names)[[1]]
  sidron_filtered&lt;-sidron[sidron_index]
  write.fasta(paste(sidron_filtered)[1], paste0(&quot;sidron_&quot;, gene_name),file.out = combined_fasta_name, open=&quot;a&quot;, as.string=TRUE)
  
  
  
  #Part 2: align the sequences using multiple sequence alignment 
  sequences &lt;- Biostrings::readAAStringSet(combined_fasta_name)
  msa &lt;-msa::msa(
  inputSeqs=sequences,
  method = &quot;ClustalW&quot;
)


return(msa)
}</code></pre>
</div>
</div>
<div id="function-that-helps-find-the-missense-variant-from-the-msa-object" class="section level2">
<h2>Function that helps find the missense variant from the msa object</h2>
<p>Additional computations in R were used to find the positions of missense variants excluding any that were the result of X’s in the sequence.</p>
<pre class="r"><code>#why is index of each sequence needed? The msa before can arrange the sequences not necessarily in the order inputted
missense_finder&lt;-function(completed_msa,human_index, gorilla_index, chimp_index, altai_index, vindija_index, sidron_index){
  
  #convert the msa to another format so that I can work with it  
  convert_msa &lt;- msaConvert(completed_msa,&quot;seqinr::alignment&quot;)
  
  missense_list&lt;-c()

  #go through entire alignment 
  for (i in 1:nchar(convert_msa$seq[1])){
    
    #if the human amino acid doesn&#39;t match the altai, add to the stored list 
    if ((substr(convert_msa$seq[human_index],i,i)!=substr(convert_msa$seq[altai_index],i,i))&amp;&amp;(substr(convert_msa$seq[altai_index],i,i)!=&quot;X&quot;)&amp;&amp;(substr(convert_msa$seq[human_index],i,i)!=&quot;-&quot;)&amp;&amp;(substr(convert_msa$seq[altai_index],i,i)!=&quot;-&quot;)){
      
      #stored in the order human, gorilla, chimp, altai, vindija, sidron
      amino_list&lt;-c(i,substr(convert_msa$seq[human_index],i,i),substr(convert_msa$seq[gorilla_index],i,i),substr(convert_msa$seq[chimp_index],i,i),substr(convert_msa$seq[altai_index],i,i),substr(convert_msa$seq[vindija_index],i,i),substr(convert_msa$seq[sidron_index],i,i))
      missense_list=rbind(missense_list, amino_list)
    }
    
    #if human amino acid doesn&#39;t match the vindija, add to the stored list 
    else if ((substr(convert_msa$seq[human_index],i,i)!=substr(convert_msa$seq[vindija_index],i,i))&amp;&amp;(substr(convert_msa$seq[vindija_index],i,i)!=&quot;X&quot;)&amp;&amp;(substr(convert_msa$seq[human_index],i,i)!=&quot;-&quot;)&amp;&amp;(substr(convert_msa$seq[vindija_index],i,i)!=&quot;-&quot;)){
      #stored in the order human, gorilla, chimp, altai, vindija, sidron
      amino_list&lt;-c(i,substr(convert_msa$seq[human_index],i,i),substr(convert_msa$seq[gorilla_index],i,i),substr(convert_msa$seq[chimp_index],i,i),substr(convert_msa$seq[altai_index],i,i),substr(convert_msa$seq[vindija_index],i,i),substr(convert_msa$seq[sidron_index],i,i))
      missense_list=rbind(missense_list, amino_list)
      
    }
    
        #if human amino acid doesn&#39;t match the sidron, add to the stored list 
    else if ((substr(convert_msa$seq[human_index],i,i)!=substr(convert_msa$seq[sidron_index],i,i))&amp;&amp;(substr(convert_msa$seq[sidron_index],i,i)!=&quot;X&quot;)&amp;&amp;(substr(convert_msa$seq[human_index],i,i)!=&quot;-&quot;)&amp;&amp;(substr(convert_msa$seq[sidron_index],i,i)!=&quot;-&quot;)){
      #stored in the order human, gorilla, chimp, altai, vindija, sidron
      amino_list&lt;-c(i,substr(convert_msa$seq[human_index],i,i),substr(convert_msa$seq[gorilla_index],i,i),substr(convert_msa$seq[chimp_index],i,i),substr(convert_msa$seq[altai_index],i,i),substr(convert_msa$seq[vindija_index],i,i),substr(convert_msa$seq[sidron_index],i,i))
      missense_list=rbind(missense_list, amino_list)
    }
    
  }

  colnames(missense_list)&lt;- c(&quot;position&quot;,&quot;human&quot;,&quot;gorilla&quot;,&quot;chimpanzee&quot;,&quot;altai&quot;,&quot;vindija&quot;, &quot;sidron&quot;)
  return(missense_list)
}</code></pre>
</div>
<div id="find-the-tooth-proteins-example" class="section level2">
<h2>Find the tooth proteins: Example</h2>
<div id="dentin-matrix-acidic-phosphoprotein-1-dmp1-ensg00000152592" class="section level4">
<h4>Dentin matrix acidic phosphoprotein 1 (DMP1): ENSG00000152592</h4>
<pre class="r"><code>dmp1_msa&lt;-msa_maker(&quot;dmp1&quot;, &quot;ENSG00000152592&quot;, &quot;uniprot_dmp1.fasta&quot;, &quot;combined_dmp1.fasta&quot;)</code></pre>
<pre><code>## use default substitution matrix</code></pre>
<pre class="r"><code>#pretty print the aligned sequence in a pdf file using the following line
#msaPrettyPrint(dmp1_msa,file=paste0(&quot;dmp1&quot;,&quot;.pdf&quot;),askForOverwrite=FALSE, showLogo=&quot;none&quot;,showConsensus = &quot;none&quot;)

missense_finder(dmp1_msa, 3, 1, 2, 4, 5, 6)</code></pre>
<pre><code>##            position human gorilla chimpanzee altai vindija sidron
## amino_list &quot;483&quot;    &quot;N&quot;   &quot;N&quot;     &quot;N&quot;        &quot;T&quot;   &quot;T&quot;     &quot;X&quot;</code></pre>
<p><img src="images/featured.png" /></p>
</div>
</div>
