---
title: "Working with Genetic Data"
author: "Cara Yijin Zou"
date: "2020-05-10"
output: html_document
draft: true 
tags: ["Bioinformatics"]
summary: "What do all the acronyms like 'NG_', 'NM_', 'ENSG', etc. mean? What are all these different files (FASTA, VCF, etc.) used for? "
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
*** As I learn more, I will hopefully continue to add to this page for things that confused me or I thought would be helpful to compile together. 

## Reference Sequence Number: What do all of these acronyms mean?
When I first started working on genetic data and especially with my [PAX9 project](../../PAX9_RF.Rmd), I was so confused by all of the different values that were associated with the gene: 

Genomic reference: NG_013357.1 (GRCh 37) 
Transcript reference: NM_006194.3
Ensembl gene ID: ENSG00000198807.8
Canonical transcript: ENST00000361487.6

## GRCh 37 
Before going into what all of these sequences are, first of all, what is GRCh37? I kept coming across this when I first trying to find data on my gene. **GRCh37** stands for Genome Reference Consortium Human Build 37 and is generally synonymous with **hg19**. Similarly, **hg38** can also be used to refer to the same genome build as **GRCh38**. 

What is the difference between GRCh37 and GRCh38? GRCh38 was released after GRCh37 and has more updated annotations. Luckily, various tools such as from [NCBI]](https://www.ncbi.nlm.nih.gov/genome/tools/remap) or [Ensembl](https://www.ncbi.nlm.nih.gov/genome/tools/remap) can be used for converting annotations. 


## NCBI Reference Sequence (RefSeq)
The accession number format: 

* NG_: gene/genomic region

* NC_: plus (+) strand on **c**hromosome 

* NT_: constructed genomic contigs (overlapping DNA fragments that are used to assemble the full sequence)

_'NC', 'NG', 'NT' all refer to genomic sequences_

* NM_: **m**RNA, coding strand 

* NR_: non-coding **R**NA 

* NP_: **p**rotein

_Also, it it begins with an 'X' liks 'XP' or 'XM', it probably means that the sequence is from computational prediction._

_What about CDS? That is the coding region of a gene that is translated to protein. Therefore, it excludes the UTR and introns._


## Ensemble 

* ENST: transcipt 

* ENSG: gene _(One gene can many corresponding transcripts.)_

* ENSP: protein 



# Types of Files  
## FASTA 
FASTA is a type of text-based file that is used to store a nucleotide or amino acid sequence. In R, there is the read.fasta() function from the 'seqinr' package. It's written in the following format 

> \>3LWB:A|PDBID|CHAIN|SEQUENCE
MSANDRRDRRVRVAVVFGGRSNEHAISCVSAGSILRNLDSRRFDVIAVGITPAGSWVLTDANPDALTITNRELPQVKSGS
GTELALPADPRRGGQLVSLPPGAGEVLESVDVVFPVLHGPYGEDGTIQGLLELAGVPYVGAGVLASAVGMDKEFTKKLLA
ADGLPVGAYAVLRPPRSTLHRQECERLGLPVFVKPARGGSSIGVSRVSSWDQLPAAVARARRHDPKVIVEAAISGRELEC
GVLEMPDGTLEASTLGEIRVAGVRGREDSFYDFATKYLDDAAELDVPAKVDDQVAEAIRQLAIRAFAAIDCRGLARVDFF
LTDDGPVINEINTMPGFTTISMYPRMWAASGVDYPTLLATMIETTLARGVGLH
\>3LWB:B|PDBID|CHAIN|SEQUENCE
MSANDRRDRRVRVAVVFGGRSNEHAISCVSAGSILRNLDSRRFDVIAVGITPAGSWVLTDANPDALTITNRELPQVKSGS
GTELALPADPRRGGQLVSLPPGAGEVLESVDVVFPVLHGPYGEDGTIQGLLELAGVPYVGAGVLASAVGMDKEFTKKLLA
ADGLPVGAYAVLRPPRSTLHRQECERLGLPVFVKPARGGSSIGVSRVSSWDQLPAAVARARRHDPKVIVEAAISGRELEC
GVLEMPDGTLEASTLGEIRVAGVRGREDSFYDFATKYLDDAAELDVPAKVDDQVAEAIRQLAIRAFAAIDCRGLARVDFF
LTDDGPVINEINTMPGFTTISMYPRMWAASGVDYPTLLATMIETTLARGVGLH

The first line is used to describe the folowing sequence and it is indicated by the ">" sign. As you can see above, multiple sequences can be found in one FASTA file. 

## Fastq
Similar to FASTA, except from DNA sequencing (commonly for illumina). It also includes the PHRED score which is a measure of the quality of the reading for each base. It is denoted by a single ASCII character. 

> @ SequenceID
> CATGGGCAGCCGAGAGATTGCGA
> +
> K<=gux;YZ[bcs3^_`a;<\b|}

The first line contains the description like fasta but is indicated by the "@" sign instead. After which, the nucleotide sequence is shown on the 2nd line. The third line starts with a "+" sign. Finally, the 4th line contains the quality scores. 

## SAM/BAM files 
SAM and BAM files contain the same information, the difference is that BAM files are **binary** versions of a SAM file. SAM stands for "Sequence Alignment/Map" and contain information about sequences aligned to a reference sequence, and is generated by next generation sequencing. 


## VCF
VCF stands for variant call format and stores different variants of a gene.


What are the different BLASTS?
BLOSUM? 

For more acronyms, I found this [glossary](http://www.nslc.wustl.edu/courses/archives/Bio3050/Glossary.pdf) online that has a lot of the basics. 

References 
1. http://useast.ensembl.org/info/website/tutorials/grch37.html

2. https://bitesizebio.com/38335/get-to-know-your-reference-genome-grch37-vs-grch38/

3. https://www.ncbi.nlm.nih.gov/books/NBK50679/#RefSeqFAQ.what_is_a_reference_sequence_r

4. https://thesequencingcenter.com/knowledge-base/fastq-files/