---
title: "Bioinformatics Glossary"
author: "Cara Yijin Zou"
date: "2020-01-10"
categories: ["Bioinformatics"]
tags: ["Bioinformatics"]
summary: "What do all the acronyms like 'NG_', 'NM_', 'ENSG', etc. mean? What are all these different files (FASTA, VCF, etc.) used for? "
output:
  blogdown::html_page:
    toc: true
---


<div id="TOC">
<ul>
<li><a href="#genetic-data">Genetic Data</a><ul>
<li><a href="#reference-sequence-number">Reference Sequence Number:</a></li>
<li><a href="#grch-37">GRCh 37</a></li>
<li><a href="#ncbi-reference-sequence-refseq">NCBI Reference Sequence (RefSeq)</a></li>
<li><a href="#ensemble">Ensemble</a></li>
</ul></li>
<li><a href="#types-of-files">Types of Files</a><ul>
<li><a href="#fasta">FASTA</a></li>
<li><a href="#fastq">Fastq</a></li>
<li><a href="#sambam-files">SAM/BAM files</a></li>
<li><a href="#vcf">VCF</a></li>
</ul></li>
<li><a href="#basic-programming-cheatsheet">Basic Programming CheatSheet</a><ul>
<li><a href="#unixlinux-bash">Unix/Linux (Bash)</a><ul>
<li><a href="#need-to-kill-a-process">Need to kill a process?</a></li>
<li><a href="#bash-script">Bash Script</a></li>
</ul></li>
<li><a href="#r-vs-python">R vs Python</a></li>
</ul></li>
<li><a href="#micellaneous">Micellaneous</a><ul>
<li><a href="#k-mers">K-mers</a></li>
<li><a href="#blast">BLAST</a></li>
<li><a href="#needleman-wunsch-algorithm">Needleman-Wunsch algorithm</a></li>
<li><a href="#smith-waterman-algorithm">Smith-Waterman algorithm</a></li>
<li><a href="#blosum-and-pam">BLOSUM and PAM</a></li>
<li><a href="#other">Other</a></li>
</ul></li>
</ul>
</div>

<p>*** As I learn more, I will hopefully continue to add to this page for things that confused me or I thought would be helpful to compile together. As a warning, this is just a collection of notes and is not super organized.</p>
<div id="genetic-data" class="section level1">
<h1>Genetic Data</h1>
<div id="reference-sequence-number" class="section level2">
<h2>Reference Sequence Number:</h2>
<p>When I first started working on genetic data and especially with my <a href="../../project/PAX9_RF">PAX9 project</a>, I was so confused by all of the different values that were associated with the gene:</p>
<p>Genomic reference: NG_013357.1 (GRCh 37)</p>
<p>Transcript reference: NM_006194.3</p>
<p>Ensembl gene ID: ENSG00000198807.8</p>
<p>Canonical transcript: ENST00000361487.6</p>
</div>
<div id="grch-37" class="section level2">
<h2>GRCh 37</h2>
<p>Before going into what all of these sequences are, first of all, what is GRCh37? I kept coming across this when I first trying to find data on my gene. <strong>GRCh37</strong> stands for Genome Reference Consortium Human Build 37 and is generally synonymous with <strong>hg19</strong>. Similarly, <strong>hg38</strong> can also be used to refer to the same genome build as <strong>GRCh38</strong>.</p>
<p>What is the difference between GRCh37 and GRCh38? GRCh38 was released after GRCh37 and has more updated annotations. Luckily, various tools such as from [NCBI]](<a href="https://www.ncbi.nlm.nih.gov/genome/tools/remap" class="uri">https://www.ncbi.nlm.nih.gov/genome/tools/remap</a>) or <a href="https://www.ncbi.nlm.nih.gov/genome/tools/remap">Ensembl</a> can be used for converting annotations.</p>
</div>
<div id="ncbi-reference-sequence-refseq" class="section level2">
<h2>NCBI Reference Sequence (RefSeq)</h2>
<p>The accession number format:</p>
<ul>
<li><p>NG_: <strong>g</strong>ene/<strong>g</strong>enomic region</p></li>
<li><p>NC_: plus (+) strand on <strong>c</strong>hromosome</p></li>
<li><p>NT_: constructed genomic contigs (overlapping DNA fragments that are used to assemble the full sequence)</p></li>
</ul>
<p><em>‘NC’, ‘NG’, ‘NT’ all refer to genomic sequences</em></p>
<ul>
<li><p>NM_: <strong>m</strong>RNA, coding strand</p></li>
<li><p>NR_: non-coding <strong>R</strong>NA</p></li>
<li><p>NP_: <strong>p</strong>rotein</p></li>
</ul>
<p><em>Also, if it begins with an ‘X’ liks ‘XP’ or ‘XM’, it probably means that the sequence is from computational prediction.</em></p>
<p><em>What about CDS? That is the coding region of a gene that is translated to protein. Therefore, it excludes the UTR and introns.</em></p>
</div>
<div id="ensemble" class="section level2">
<h2>Ensemble</h2>
<ul>
<li><p>ENST: transcipt</p></li>
<li><p>ENSG: gene <em>(One gene can have many corresponding transcripts.)</em></p></li>
<li><p>ENSP: protein</p></li>
</ul>
</div>
</div>
<div id="types-of-files" class="section level1">
<h1>Types of Files</h1>
<div id="fasta" class="section level2">
<h2>FASTA</h2>
<p>FASTA is a type of text-based file that is used to store a nucleotide or amino acid sequence. In R, there is the read.fasta() function from the ‘seqinr’ package. It’s written in the following format:</p>
<blockquote>
<p>&gt;3LWB:A|PDBID|CHAIN|SEQUENCE
MSANDRRDRRVRVAVVFGGRSNEHAISCVSAGSILRNLDSRRFDVIAVGITPAGSWVLTDANPDALTITNRELPQVKSGS
GTELALPADPRRGGQLVSLPPGAGEVLESVDVVFPVLHGPYGEDGTIQGLLELAGVPYVGAGVLASAVGMDKEFTKKLLA
ADGLPVGAYAVLRPPRSTLHRQECERLGLPVFVKPARGGSSIGVSRVSSWDQLPAAVARARRHDPKVIVEAAISGRELEC
GVLEMPDGTLEASTLGEIRVAGVRGREDSFYDFATKYLDDAAELDVPAKVDDQVAEAIRQLAIRAFAAIDCRGLARVDFF
LTDDGPVINEINTMPGFTTISMYPRMWAASGVDYPTLLATMIETTLARGVGLH
&gt;3LWB:B|PDBID|CHAIN|SEQUENCE
MSANDRRDRRVRVAVVFGGRSNEHAISCVSAGSILRNLDSRRFDVIAVGITPAGSWVLTDANPDALTITNRELPQVKSGS
GTELALPADPRRGGQLVSLPPGAGEVLESVDVVFPVLHGPYGEDGTIQGLLELAGVPYVGAGVLASAVGMDKEFTKKLLA
ADGLPVGAYAVLRPPRSTLHRQECERLGLPVFVKPARGGSSIGVSRVSSWDQLPAAVARARRHDPKVIVEAAISGRELEC
GVLEMPDGTLEASTLGEIRVAGVRGREDSFYDFATKYLDDAAELDVPAKVDDQVAEAIRQLAIRAFAAIDCRGLARVDFF
LTDDGPVINEINTMPGFTTISMYPRMWAASGVDYPTLLATMIETTLARGVGLH</p>
</blockquote>
<p>The first line is used to describe the folowing sequence and it is indicated by the “&gt;” sign. As you can see above, multiple sequences can be found in one FASTA file.</p>
</div>
<div id="fastq" class="section level2">
<h2>Fastq</h2>
<p>Similar to FASTA, except from DNA sequencing (commonly for illumina). It also includes the PHRED score which is a measure of the quality of the reading for each base. It is denoted by a single ASCII character.</p>
<blockquote>
<p>@ SequenceID</p>
</blockquote>
<blockquote>
<p>CATGGGCAGCCGAGAGATTGCGA</p>
</blockquote>
<blockquote>
<p>+</p>
</blockquote>
<blockquote>
<p>K&lt;=gux;YZ[bcs3^_`a;&lt;|}</p>
</blockquote>
<p>The first line contains the description like fasta but is indicated by the “@” sign instead. After which, the nucleotide sequence is shown on the 2nd line. The third line starts with a “+” sign. Finally, the 4th line contains the quality scores.</p>
</div>
<div id="sambam-files" class="section level2">
<h2>SAM/BAM files</h2>
<p>SAM and BAM files contain the same information, the difference is that BAM files are <strong>binary</strong> versions of a SAM file. SAM stands for “Sequence Alignment/Map” and contain information about sequences aligned to a reference sequence, and is generated by next generation sequencing.</p>
</div>
<div id="vcf" class="section level2">
<h2>VCF</h2>
<p>VCF stands for variant call format and stores different variants of a gene.</p>
</div>
</div>
<div id="basic-programming-cheatsheet" class="section level1">
<h1>Basic Programming CheatSheet</h1>
<div id="unixlinux-bash" class="section level2">
<h2>Unix/Linux (Bash)</h2>
<p>Another way of navigating through files and giving commands to the operating system.</p>
<p><strong>Absolute</strong> path (begins with “/”)</p>
<blockquote>
<p>/home/Downloads/pax9.csv</p>
</blockquote>
<p><strong>Relative</strong> path (does NOT begin with “/”)</p>
<p>If you are in /home</p>
<blockquote>
<p>Downloads</p>
</blockquote>
<blockquote>
<p>Downloads/pax9.csv</p>
</blockquote>
<p>If you are in /home/Downloads</p>
<blockquote>
<p>pax9.csv</p>
</blockquote>
<table>
<colgroup>
<col width="31%" />
<col width="68%" />
</colgroup>
<thead>
<tr class="header">
<th>Command</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pwd</td>
<td>“print working directory” (absolute)</td>
</tr>
<tr class="even">
<td>/</td>
<td>root directory</td>
</tr>
<tr class="odd">
<td>ls</td>
<td>“listing”, gives contents of current directory</td>
</tr>
<tr class="even">
<td><strong>ls</strong> /home/Downloads</td>
<td>contents of download directory</td>
</tr>
<tr class="odd">
<td>cd</td>
<td>“change directory”</td>
</tr>
<tr class="even">
<td>..</td>
<td>directory above current</td>
</tr>
<tr class="odd">
<td>.</td>
<td>current directory</td>
</tr>
<tr class="even">
<td>~</td>
<td>home directory</td>
</tr>
<tr class="odd">
<td>q</td>
<td>“quit”</td>
</tr>
<tr class="even">
<td>Control + C</td>
<td>“cancel”</td>
</tr>
<tr class="odd">
<td>top</td>
<td>allows you to view the current processes running</td>
</tr>
<tr class="even">
<td><strong>cp</strong> test.csv test2.csv</td>
<td>“copy”: test.csv is duplicated and named test2.csv (last arg = destination)</td>
</tr>
<tr class="odd">
<td><strong>cp</strong> test.csv other.csv Downloads</td>
<td>“copy”: test.csv and other.csv is copied to the Downloads directory</td>
</tr>
<tr class="even">
<td><strong>mv</strong> test.csv other.csv Downloads</td>
<td>“moves”: test.csv and other.csv is moved to the downloads directory</td>
</tr>
<tr class="odd">
<td><strong>mv</strong> test.csv new.csv</td>
<td>“renames”: test.csv is renamed to new.csv (also works for directory)</td>
</tr>
<tr class="even">
<td><strong>rm</strong> test.csv other.csv</td>
<td>“removes”: deletes other.csv and test.csv (does not work for directory)</td>
</tr>
<tr class="odd">
<td><strong>rmdir</strong> Downloads</td>
<td>“removes”: directory (must be empty!)</td>
</tr>
<tr class="even">
<td><strong>mkdir</strong> Homework</td>
<td>“makes directory” called homework</td>
</tr>
<tr class="odd">
<td><strong>man</strong> ls</td>
<td>“manual” of command ls</td>
</tr>
<tr class="even">
<td><strong>chmod</strong> u+x test.sh</td>
<td>“change file mode”: permissions, read ( r ), write ( w ) or execute ( x )</td>
</tr>
</tbody>
</table>
<p>Useful Tricks:</p>
<ul>
<li><p>Hit up and down arrow keys to get previous commands</p></li>
<li><p>Use tab key for autocompletion</p></li>
<li><p>Spaces for file names can cause problems because they are seen as separate items. To prevent this, put them in quotes or use  before the space</p></li>
<li><p>if a file/destination does not exist, it will create one. If it does exist, it may overwrite</p></li>
<li><p>there is no <strong>undo</strong></p></li>
</ul>
<div id="need-to-kill-a-process" class="section level3">
<h3>Need to kill a process?</h3>
<ol style="list-style-type: decimal">
<li><p>Get PID (process ID) from top</p></li>
<li><p>type: kill PID</p></li>
<li><p>doesn’t work? type: kill -9 PID</p></li>
</ol>
</div>
<div id="bash-script" class="section level3">
<h3>Bash Script</h3>
<p>A text file with commands. Anything you put in command line can be in a script and vice versa. Uses the <strong>.sh</strong> extension.</p>
<blockquote>
<p>#!/bin/bash</p>
</blockquote>
<p>First line must always look something like this. “#!” is called a <strong>shebang</strong> and immediately after (no spaces), put the path to the interpreter. If you don’t know, type: “which bash” in command line.</p>
<blockquote>
<p>#running program from its path</p>
</blockquote>
<blockquote>
<p>#Bob 07/01/2020</p>
</blockquote>
<p>Comments to describe what it does, author, date, etc.</p>
<blockquote>
<p>name = “Bob”</p>
</blockquote>
<p>You can set variables</p>
<blockquote>
<p>echo Hello $name !</p>
</blockquote>
<p>Like print, it will print stuff after it. Refer to variables with a “$” in front.</p>
<blockquote>
<p>pwd</p>
</blockquote>
<p>You can put commands in the script.</p>
<p>In the actual command line, to <strong>run</strong> the script:</p>
<blockquote>
<p>./testscript.sh</p>
</blockquote>
<p>AKA, look in the current directory to find the script named testscript.sh and run it.</p>
</div>
</div>
<div id="r-vs-python" class="section level2">
<h2>R vs Python</h2>
<p>While R has a lot of super useful packages, especially for bioinformatics and statistics, I found it <strong>super</strong> annoying to code in base R compared to python. Here are some essential equivalents for R:</p>
<table>
<colgroup>
<col width="30%" />
<col width="69%" />
</colgroup>
<thead>
<tr class="header">
<th>Python</th>
<th>R</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>len(myList)</td>
<td>length(myList) **won’t work for strings (see below example)</td>
</tr>
<tr class="even">
<td>x in myList</td>
<td>x %in% myList</td>
</tr>
<tr class="odd">
<td>myList.index(item)</td>
<td>match (item, myList)</td>
</tr>
<tr class="even">
<td>“string”[:3]</td>
<td>substr (“string”, 1, 3)</td>
</tr>
<tr class="odd">
<td>range(0, 6,2)</td>
<td>seq(from = 0, 4, by=2)</td>
</tr>
<tr class="even">
<td>“str”+“ing”</td>
<td>paste0(“str”,“ing”, &quot;&quot;)</td>
</tr>
<tr class="odd">
<td>myList=[1,2]</td>
<td>myList=c(1,2)</td>
</tr>
<tr class="even">
<td>myList.append(3)</td>
<td>myList=c(myList, 3)</td>
</tr>
<tr class="odd">
<td>myList.extend(myList2)</td>
<td>myList=c(myList,myList2) <strong>OR</strong> myList=append(myList, myList2)</td>
</tr>
</tbody>
</table>
<p><strong>Examples</strong></p>
<ol style="list-style-type: decimal">
<li><strong>python</strong>: len(“string”) vs <strong>R</strong>: nchar(“string”)</li>
</ol>
<pre class="python"><code>for i in range(len(&quot;test&quot;)): 
  print (i)</code></pre>
<pre><code>## 0
## 1
## 2
## 3</code></pre>
<pre class="r"><code>for (i in 1:nchar(&quot;test&quot;)){
  print (i)
}</code></pre>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Matrices</li>
</ol>
<p><strong>python</strong></p>
<pre class="python"><code>test_matrix=[]
for i in range(3): 
  test_matrix.append([0]*3)
test_matrix[0][1]=2
print(test_matrix)</code></pre>
<pre><code>## [[0, 2, 0], [0, 0, 0], [0, 0, 0]]</code></pre>
<p><strong>R</strong></p>
<pre class="r"><code>test_matrix&lt;- matrix(0, nrow = 3, ncol = 3)
test_matrix[1,2]=2
print(test_matrix)</code></pre>
<pre><code>##      [,1] [,2] [,3]
## [1,]    0    2    0
## [2,]    0    0    0
## [3,]    0    0    0</code></pre>
</div>
</div>
<div id="micellaneous" class="section level1">
<h1>Micellaneous</h1>
<div id="k-mers" class="section level2">
<h2>K-mers</h2>
<p>Subsequences of length k of a nucleotide sequence.</p>
<pre class="python"><code>def findkmers(k, seq):
    dict1={}
    a = 0 
    b= k
    k_mer= seq[a:b]
    while len(k_mer)==k: 
        if k_mer in dict1: 
            dict1[k_mer]+= 1
        else: 
            dict1[k_mer]= 1
        a+=1
        b+=1
        k_mer= seq[a:b]
    print (dict1)

def main():
   my_seq1 = &quot;CAGCCCAATC&quot;
   print(&quot;For the sequence: &quot;, my_seq1)
   for i in range (1,5):
      print(&quot;The&quot;, str(i)+&quot;-mers are :&quot;)
      findkmers(i, my_seq1)

main()</code></pre>
<pre><code>## For the sequence:  CAGCCCAATC
## The 1-mers are :
## {&#39;C&#39;: 5, &#39;A&#39;: 3, &#39;G&#39;: 1, &#39;T&#39;: 1}
## The 2-mers are :
## {&#39;CA&#39;: 2, &#39;AG&#39;: 1, &#39;GC&#39;: 1, &#39;CC&#39;: 2, &#39;AA&#39;: 1, &#39;AT&#39;: 1, &#39;TC&#39;: 1}
## The 3-mers are :
## {&#39;CAG&#39;: 1, &#39;AGC&#39;: 1, &#39;GCC&#39;: 1, &#39;CCC&#39;: 1, &#39;CCA&#39;: 1, &#39;CAA&#39;: 1, &#39;AAT&#39;: 1, &#39;ATC&#39;: 1}
## The 4-mers are :
## {&#39;CAGC&#39;: 1, &#39;AGCC&#39;: 1, &#39;GCCC&#39;: 1, &#39;CCCA&#39;: 1, &#39;CCAA&#39;: 1, &#39;CAAT&#39;: 1, &#39;AATC&#39;: 1}</code></pre>
</div>
<div id="blast" class="section level2">
<h2>BLAST</h2>
<p><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> stands for Basic Local Alignment Search Tool. It finds areas of similarity between sequences of nucleotides or amino acids. There are many types of blast including blastn, blastp, blastx, tblastn, etc.</p>
<ul>
<li><p>Query cover: how much of the query sequence actually overlapps with the aligned sequence.</p></li>
<li><p>E value (expected value): smaller is better (significant). It basically means how many hits would you get just randomly (based on the quality of score and the length of the query)</p></li>
<li><p>Percent Identity: percentage of characters matching in the query and aligned sequence</p></li>
</ul>
</div>
<div id="needleman-wunsch-algorithm" class="section level2">
<h2>Needleman-Wunsch algorithm</h2>
<p>Global alignment</p>
</div>
<div id="smith-waterman-algorithm" class="section level2">
<h2>Smith-Waterman algorithm</h2>
<p>Local alignment</p>
</div>
<div id="blosum-and-pam" class="section level2">
<h2>BLOSUM and PAM</h2>
<p>Both used as a measure of similarity between proteins. PAM is for closely related while BLOSUM is for distantly related. For BLOSUM, BLOSUM45 would be used for proteins more distant from each other than BLOSUM90.</p>
</div>
<div id="other" class="section level2">
<h2>Other</h2>
<p>For more acronyms, I found this <a href="http://www.nslc.wustl.edu/courses/archives/Bio3050/Glossary.pdf">glossary</a> online that has a lot of the basics.</p>
<p>References</p>
<ol style="list-style-type: decimal">
<li><p><a href="http://useast.ensembl.org/info/website/tutorials/grch37.html" class="uri">http://useast.ensembl.org/info/website/tutorials/grch37.html</a></p></li>
<li><p><a href="https://bitesizebio.com/38335/get-to-know-your-reference-genome-grch37-vs-grch38/" class="uri">https://bitesizebio.com/38335/get-to-know-your-reference-genome-grch37-vs-grch38/</a></p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/books/NBK50679/#RefSeqFAQ.what_is_a_reference_sequence_r" class="uri">https://www.ncbi.nlm.nih.gov/books/NBK50679/#RefSeqFAQ.what_is_a_reference_sequence_r</a></p></li>
<li><p><a href="https://thesequencingcenter.com/knowledge-base/fastq-files/" class="uri">https://thesequencingcenter.com/knowledge-base/fastq-files/</a></p></li>
<li><p>Kerfeld and Scott, PLoS Biology 2011</p></li>
<li><p><a href="https://ryanstutorials.net/linuxtutorial/" class="uri">https://ryanstutorials.net/linuxtutorial/</a></p></li>
<li><p><a href="https://ryanstutorials.net/bash-scripting-tutorial/bash-script.php" class="uri">https://ryanstutorials.net/bash-scripting-tutorial/bash-script.php</a></p></li>
<li><p><a href="https://learn.datacamp.com/" class="uri">https://learn.datacamp.com/</a></p></li>
</ol>
</div>
</div>
